Source/WebCore/rendering/svg/SVGResources.cpp

```diff
@@ -313,23 +313,20 @@ bool SVGResources::buildCachedResources(const RenderElement& renderer, const Ren
 
 void SVGResources::layoutDifferentRootIfNeeded(const LegacyRenderSVGRoot* svgRoot)
 {
-    if (clipper() && svgRoot != SVGRenderSupport::findTreeRootObject(*clipper()))
-        clipper()->layoutIfNeeded();
-
-    if (masker() && svgRoot != SVGRenderSupport::findTreeRootObject(*masker()))
-        masker()->layoutIfNeeded();
-
-    if (filter() && svgRoot != SVGRenderSupport::findTreeRootObject(*filter()))
-        filter()->layoutIfNeeded();
-
-    if (markerStart() && svgRoot != SVGRenderSupport::findTreeRootObject(*markerStart()))
-        markerStart()->layoutIfNeeded();
-
-    if (markerMid() && svgRoot != SVGRenderSupport::findTreeRootObject(*markerMid()))
-        markerMid()->layoutIfNeeded();
-
-    if (markerEnd() && svgRoot != SVGRenderSupport::findTreeRootObject(*markerEnd()))
-        markerEnd()->layoutIfNeeded();
+    auto layoutDifferentRootIfNeeded = [&](RenderElement* container) {
+        if (!container)
+            return;
+        auto* root = SVGRenderSupport::findTreeRootObject(*container);
+        if (svgRoot == root || root->isInLayout())
+            return;
```

Comment:
Why do we need to add this check? Does not SVGResourcesCycleSolver::resolveCycles() break the cycle between the SVG resources?

Reply:
SVGResourcesCycleSolver::resolveCycles DOES break the cycle. What we need to prevent here is the (mutual) recursion of SVGRenderSupport::layoutDifferentRootIfNeeded -> SVGResources::layoutDifferentRootIfNeeded -> RenderElement::layoutIfNeeded -> 
RenderSVGResourceContainer::layout -> LegacyRenderSVGHiddenContainer::layout -> SVGRenderSupport::layoutChildren -> SVGRenderSupport::layoutDifferentRootIfNeeded. This recursion exists independent of SVG resource cycles (when there are more than two SVG root elements whose descendant mutually references each other) so we need to manually prevent infinite recursion here.